lib/ddtrace/sampler.rb

```diff
@@ -0,0 +1,48 @@
+
+module Datadog
+  # \Sampler performs client-side trace sampling.
+  class Sampler
+    def sample(span); end
+  end
+
+  # \AllSampler samples all the traces.
+  class AllSampler < Sampler
+    def sample(span)
+      span.sampled = true
+    end
+  end
+
+  # \RateSampler is based on a sample rate.
+  class RateSampler < Sampler
+    KNUTH_FACTOR = 1111111111111111111
+    SAMPLE_RATE_METRIC_KEY = '_sample_rate'.freeze()
+
+    attr_reader :sample_rate
+
+    # Initialize a \RateSampler.
+    # This sampler keeps a random subset of the traces. Its main purpose is to
+    # reduce the instrumentation footprint.
+    #
+    # * +sample_rate+: the sample rate as a \Float between 0.0 and 1.0. 0.0
+    #   means that no trace will be sampled; 1.0 means that all traces will be
+    #   sampled.
+    def initialize(sample_rate = 1.0)
+      unless sample_rate > 0.0 && sample_rate <= 1.0
+        raise ArgumentError,
```

Comment:
As I already explained, raising an exception is not a crash. `ArgumentError` exceptions are designed very precisely for this kind of problem. Logging errors and disabling the sampler is very bad because people miss errors, and then believe the code works as they intended to... even though it does not. If they use an invalid sample rate, the code will raise an exception, which they won't be able to miss, and they will be able to fix it before going to  prod. In C you would use an assertion, with the exact same purpose.